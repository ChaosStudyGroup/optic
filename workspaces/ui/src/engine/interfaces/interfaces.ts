import { IDiff } from './diffs';

interface SerializedDiff {
  [key: string]: {
    interactionTrail: {
      path: any[];
    };
    requestsTrail: any;
  };
}

export enum ICoreShapeKinds {
  ObjectKind = '$object',
  ListKind = '$list',
  MapKind = '$map',
  OneOfKind = '$oneOf',
  AnyKind = '$any',
  StringKind = '$string',
  NumberKind = '$number',
  BooleanKind = '$boolean',
  NullableKind = '$nullable',
  OptionalKind = '$optional',
  UnknownKind = '$unknown',
}

// Diff Types the UI Handles

export const allowedDiffTypes: {
  [key: string]: { isBodyShapeDiff: boolean; asString: string };
} = {
  UnmatchedRequestUrl: {
    isBodyShapeDiff: false,
    asString: 'UnmatchedRequestUrl',
  },
  UnmatchedRequestMethod: {
    isBodyShapeDiff: false,
    asString: 'UnmatchedRequestMethod',
  },
  UnmatchedRequestBodyContentType: {
    isBodyShapeDiff: false,
    asString: 'UnmatchedRequestBodyContentType',
  },
  UnmatchedResponseBodyContentType: {
    isBodyShapeDiff: false,
    asString: 'UnmatchedResponseBodyContentType',
  },
  UnmatchedResponseStatusCode: {
    isBodyShapeDiff: false,
    asString: 'UnmatchedResponseStatusCode',
  },
  UnmatchedRequestBodyShape: {
    isBodyShapeDiff: true,
    asString: 'UnmatchedRequestBodyShape',
  },
  UnmatchedResponseBodyShape: {
    isBodyShapeDiff: true,
    asString: 'UnmatchedResponseBodyShape',
  },
};

export const allowedDiffTypesKeys: string[] = Object.keys(allowedDiffTypes);

// Properties of Each Diff Types

export const isBodyShapeDiff = (key: string): boolean =>
  allowedDiffTypes[key]?.isBodyShapeDiff;
export const DiffInRequest = (key: string): boolean =>
  key === 'UnmatchedRequestBodyShape';
export const DiffInResponse = (key: string): boolean =>
  key === 'UnmatchedResponseBodyShape';

// The ones we like to work with in the UI

export interface IRequestBodyLocation {
  contentType?: string;
}

export interface IResponseBodyLocation {
  statusCode: number;
  contentType?: string;
}

export interface IParsedLocation {
  pathId: string;
  method: string;
  inRequest?: IRequestBodyLocation;
  inResponse?: IResponseBodyLocation;
}
