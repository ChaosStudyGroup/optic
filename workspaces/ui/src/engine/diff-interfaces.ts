interface SerializedDiff {
  [key: string]: {
    interactionTrail: {
      path: any[];
    };
    requestsTrail: any;
  };
}

// Diff Types the UI Handles

export const allowedDiffTypes: {
  [key: string]: { isBodyShapeDiff: boolean };
} = {
  UnmatchedRequestUrl: { isBodyShapeDiff: false },
  UnmatchedRequestMethod: {
    isBodyShapeDiff: false,
  },
  UnmatchedRequestBodyContentType: {
    isBodyShapeDiff: false,
  },
  UnmatchedResponseBodyContentType: {
    isBodyShapeDiff: false,
  },
  UnmatchedResponseStatusCode: {
    isBodyShapeDiff: false,
  },
  UnmatchedRequestBodyShape: {
    isBodyShapeDiff: true,
  },
  UnmatchedResponseBodyShape: {
    isBodyShapeDiff: true,
  },
};

export const allowedDiffTypesKeys: string[] = Object.keys(allowedDiffTypes);

// Properties of Each Diff Types

export const isBodyShapeDiff = (key: string): boolean =>
  allowedDiffTypes[key]?.isBodyShapeDiff;
export const DiffInRequest = (key: string): boolean =>
  key === 'UnmatchedRequestBodyShape';
export const DiffInResponse = (key: string): boolean =>
  key === 'UnmatchedResponseBodyShape';

// The ones we like to work with in the UI

export interface BodyShapeDiff<ShapeDiffLocation> {
  diff: any;
  type: string;
  interactionPointers: string[];
  location: Location;
}

export interface RequestBodyLocation {
  contentType: string;
}

export interface ResponseBodyLocation {
  statusCode: number;
  contentType: string;
}
