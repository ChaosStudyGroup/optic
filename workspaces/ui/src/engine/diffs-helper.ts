import invariant from 'invariant';
import {
  allowedDiffTypesKeys,
  BodyShapeDiff,
  DiffInRequest,
  DiffInResponse,
  isBodyShapeDiff,
  RequestBodyLocation,
} from './interfaces/interfaces';
import {
  IUnmatchedRequestBodyShape,
  IUnmatchedResponseBodyShape,
} from './interfaces/diffs';

class DiffLifecycleManager {
  constructor(private serialized_diff: any, interactions: string[]) {}

  diffType(): string {
    const keys = Object.keys(this.serialized_diff);
    const typeKey = keys[0];
    invariant(
      keys.length === 1 && allowedDiffTypesKeys.includes(typeKey),
      'Serialized diffs should only have one root key'
    );
    return typeKey!;
  }

  isShapeDiff(): boolean {
    return isBodyShapeDiff(this.diffType());
  }

  inRequest(): boolean {
    return DiffInRequest(this.diffType());
  }

  inResponse(): boolean {
    return DiffInResponse(this.diffType());
  }

  asRequestBodyShapeDiff: BodyShapeDiff<
    IUnmatchedRequestBodyShape,
    RequestBodyLocation
  > = {};

  asResponseBodyShapeDiff: BodyShapeDiff<
    IUnmatchedResponseBodyShape,
    RequestBodyLocation
  > = {};
}
