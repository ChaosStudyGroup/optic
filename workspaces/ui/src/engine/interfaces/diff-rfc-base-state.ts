import { opticEngine, Queries } from '@useoptic/domain';

export interface DiffRfcBaseState {
  eventStore: any;
  rfcState: any;
  rfcService: any;
  rfcId: string;
  queries: any;
  domainIdGenerator: {
    newShapeId: () => string;
    newPathId: () => string;
    newRequestId: () => string;
    newResponseId: () => string;
    newShapeParameterId: () => string;
    newRequestParameterId: () => string;
    newFieldId: () => string;
  };
}

export function makeDiffRfcBaseState(
  eventStore: any,
  rfcService: any,
  rfcId: string,
  domainIdGenerator: any = opticEngine.com.useoptic.OpticIdsJsHelper().random
): DiffRfcBaseState {
  const queries = Queries(eventStore, rfcService, rfcId);

  return {
    queries,
    eventStore,
    rfcService,
    rfcId,
    domainIdGenerator,
    rfcState: rfcService.currentState(rfcId),
  };
}
