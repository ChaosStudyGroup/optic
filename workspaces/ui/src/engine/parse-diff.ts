import {
  IDiff,
  IDiffWithShapeDiff,
  IShapeDiffResult,
} from './interfaces/diffs';
import jsonStringify from 'json-stable-stringify';
import sha1 from 'node-sha1';
import {
  allowedDiffTypes,
  allowedDiffTypesKeys,
  DiffInRequest,
  DiffInResponse,
  IParsedLocation,
  isBodyShapeDiff,
} from './interfaces/interfaces';
import invariant from 'invariant';
import { IShapeTrail } from './interfaces/shape-trail';
import { IInteractionTrail } from './interfaces/interaction-trail';
import { DiffRfcBaseState } from './interfaces/diff-rfc-base-state';
import { locationForTrails } from './interfaces/trail-parsers';
import { IRequestSpecTrail } from './interfaces/request-spec-trail';
import { IJsonTrail } from '@useoptic/cli-shared/build/diffs/json-trail';

export class ParsedDiff {
  diffType: string;

  constructor(
    private serialized_diff: IDiff,
    public interactions: string[],
    private rfcBaseState: DiffRfcBaseState
  ) {
    const keys = Object.keys(this.serialized_diff);
    const typeKey = keys[0]!;
    invariant(
      keys.length === 1 && allowedDiffTypesKeys.includes(typeKey),
      'Serialized diffs should only have one root key'
    );

    this.diffType = typeKey!;
  }

  raw(): IDiff {
    return this.serialized_diff;
  }

  interactionTrail(): IInteractionTrail {
    const key = Object.keys(this.serialized_diff)[0];
    return this.serialized_diff[key].interactionTrail;
  }

  requestsTrail(): IRequestSpecTrail {
    const key = Object.keys(this.serialized_diff)[0];
    return this.serialized_diff[key].requestsTrail;
  }

  location(): IParsedLocation {
    const location = locationForTrails(
      this.requestsTrail(),
      this.interactionTrail(),
      this.rfcBaseState
    );

    invariant(
      Boolean(location.pathId),
      'Diffs handled by the UI should have a known endpoint'
    );

    return {
      pathId: location!.pathId,
      method: location!.method,
      inRequest: DiffInRequest(this.diffType) && {
        contentType: location.contentType,
      },
      inResponse: DiffInResponse(this.diffType) && {
        statusCode: location.statusCode!,
        contentType: location.contentType,
      },
    };
  }

  isNewRegionDiff(): boolean {
    return !isBodyShapeDiff(this.diffType);
  }
  isBodyShapeDiff(): boolean {
    return isBodyShapeDiff(this.diffType);
  }

  isA(k: string): boolean {
    return this.diffType === k;
  }

  asShapeDiff(): BodyShapeDiff | undefined {
    if (!isBodyShapeDiff(this.diffType)) {
      return undefined;
    }

    const asWithShapeDiff = this.serialized_diff as IDiffWithShapeDiff;

    const requestBodyShapeDiff: IShapeDiffResult | undefined =
      asWithShapeDiff[allowedDiffTypes.UnmatchedRequestBodyShape.asString]
        ?.shapeDiffResult;

    const responseBodyShapeDiff: IShapeDiffResult | undefined =
      asWithShapeDiff[allowedDiffTypes.UnmatchedResponseBodyShape.asString]
        ?.shapeDiffResult;

    return new BodyShapeDiff(
      this,
      this.serialized_diff,
      (requestBodyShapeDiff || responseBodyShapeDiff)!,
      this.interactions,
      this.location()
    );
  }
}

export function parseDiffsArray(
  array: [IDiff, string[]][],
  rfcBaseState: DiffRfcBaseState
): ParsedDiff[] {
  return array.map(
    ([rawDiff, pointers]) => new ParsedDiff(rawDiff, pointers, rfcBaseState)
  );
}

export class BodyShapeDiff {
  shapeTrail: IShapeTrail;
  jsonTrail: IJsonTrail;
  shapeDiffGroupingHash: string;
  isUnmatched: boolean;
  isUnspecified: boolean;

  constructor(
    private parsedDiff: ParsedDiff,
    private diff: IDiff,
    private shapeDiff: IShapeDiffResult,
    public interactionPointers: string[],
    public location: IParsedLocation
  ) {
    this.shapeTrail = (shapeDiff['UnmatchedShape']?.shapeTrail ||
      shapeDiff['UnspecifiedShape']?.shapeTrail)!;
    invariant(
      this.shapeTrail,
      'A shape trail must be specified with all shape diffs'
    );
    this.jsonTrail = (shapeDiff['UnmatchedShape']?.jsonTrail ||
      shapeDiff['UnspecifiedShape']?.jsonTrail)!;

    this.shapeDiffGroupingHash = sha1(
      jsonStringify(this.shapeTrail),
      jsonStringify(this.jsonTrail)
    );

    this.isUnmatched = Boolean(shapeDiff['UnmatchedShape']);
    this.isUnspecified = Boolean(shapeDiff['UnspecifiedShape']);
  }
}
