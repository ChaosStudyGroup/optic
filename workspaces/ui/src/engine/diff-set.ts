import { BodyShapeDiff, ParsedDiff } from './parse-diff';
import groupby from 'lodash.groupby';
import { IShapeTrail } from './interfaces/shape-trail';
import { diff } from 'react-ace';
import { DiffRfcBaseState } from './interfaces/diff-rfc-base-state';

export class DiffSet {
  constructor(
    private diffs: ParsedDiff[],
    private rfcBaseState: DiffRfcBaseState
  ) {}

  count(): number {
    return this.diffs.length;
  }

  forEndpoint(pathId: string, method: string): DiffSet {
    return new DiffSet(
      this.diffs.filter(
        (i) =>
          i.location(this.rfcBaseState).pathId === pathId &&
          i.location(this.rfcBaseState).method === method
      ),
      this.rfcBaseState
    );
  }
  newRegions(): DiffSet {
    return new DiffSet(
      this.diffs.filter((i) => !Boolean(i.asShapeDiff(this.rfcBaseState))),
      this.rfcBaseState
    );
  }
  shapeDiffs(): DiffSet {
    return new DiffSet(
      this.diffs.filter((i) => Boolean(i.asShapeDiff(this.rfcBaseState))),
      this.rfcBaseState
    );
  }

  groupedByEndpoint(): {
    pathId: string;
    method: string;
    diffs: ParsedDiff[];
  }[] {
    const groupedByEndpoint: { [key: string]: ParsedDiff[] } = groupby(
      this.diffs,
      (d) => {
        const { pathId, method } = d.location(this.rfcBaseState);
        return `${method}.${pathId}`;
      }
    );

    return Object.entries(groupedByEndpoint).map(([key, diffs]) => {
      const { pathId, method } = diffs[0].location(this.rfcBaseState);
      return { pathId, method, diffs };
    });
  }

  groupedByEndpointAndShapeTrail(): {
    shapeDiffGroupingHash: string;
    shapeTrail: IShapeTrail;
    diffs: ParsedDiff[];
  }[] {
    const groupedByShapeTrailHash: { [key: string]: ParsedDiff[] } = groupby(
      this.shapeDiffs().iterator(),
      (i) => {
        // console.warn('make this look at endpoint too');
        return i.asShapeDiff(this.rfcBaseState)!.shapeDiffGroupingHash;
      }
    );

    return Object.entries(groupedByShapeTrailHash).map(([key, diffs]) => {
      const oneDiff = diffs[0]!.asShapeDiff(this.rfcBaseState);
      return {
        shapeDiffGroupingHash: key,
        shapeTrail: oneDiff.shapeTrail,
        jsonTrail: oneDiff.jsonTrail,
        diffs,
      };
    });
  }

  iterator(): ParsedDiff[] {
    return this.diffs;
  }
}
