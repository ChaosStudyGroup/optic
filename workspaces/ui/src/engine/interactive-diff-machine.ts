import { BodyShapeDiff, ParsedDiff } from './parse-diff';
import { assign, Machine, send, sendParent } from 'xstate';
import { IShapeTrail } from './interfaces/shape-trail';
import {
  IDiffSuggestionPreview,
  initialTitleForNewRegions,
  prepareNewRegionDiffSuggestionPreview,
  prepareShapeDiffSuggestionPreview,
} from '../__tests__/diff-helpers/prepare-diff-previews';
import { InteractiveSessionConfig } from './interfaces/session';
import { IDiffDescription } from './interfaces/interpretors';
import { descriptionForDiffs } from './interpretors/DiffDescriptionInterpretors';
import {
  ILearnedBodies,
  IValueAffordanceSerializationWithCounter,
} from '@useoptic/cli-shared/build/diffs/initial-types';

interface DiffStateSchema {
  states: {
    unfocused: {};
    loading: {};
    ready: {
      states: {
        unhandled: {};
        handled: {};
      };
    };
  };
}

// The events that the machine handles
type DiffEvent =
  | { type: 'SHOWING' }
  | { type: 'CHANGE_SUGGESTION'; to: string }
  | { type: 'DISMISS_TYPE'; type_slug: string }
  | { type: 'STAGE' }
  | { type: 'UNSTAGE' }
  | { type: 'RESET' };

// The context (extended state) of the machine
interface DiffContext<InterpretationContext> {
  results: InterpretationContext | undefined;
  preview: IDiffSuggestionPreview | undefined;
  descriptionWhileLoading: IDiffDescription;
}

export const createNewRegionMachine = (
  id: string,
  parsedDiff: ParsedDiff,
  services: InteractiveSessionConfig
) =>
  Machine<DiffContext<ILearnedBodies>, DiffStateSchema, DiffEvent>({
    id,
    context: {
      results: undefined,
      preview: undefined,
      descriptionWhileLoading: descriptionForDiffs([parsedDiff]),
    },
    initial: 'unfocused',
    states: {
      unfocused: {
        on: {
          //@ts-ignore
          'done.invoke.loading-initial-regions': {
            actions: assign({
              //@ts-ignore
              results: (context, event) => event.data,
            }),
            target: 'loading',
          },
        },
      },
      loading: {
        invoke: {
          id: 'preparing-preview-of-diff',
          src: async (context, event) => {
            return prepareNewRegionDiffSuggestionPreview(
              parsedDiff,
              services,
              context.results!
            );
          },
          onDone: {
            target: 'ready',
            actions: assign({
              preview: (context, event) => event.data,
            }),
          },
        },
      },
      ready: {
        initial: 'unhandled',
        states: {
          unhandled: {
            entry: [
              (context, event) => console.log('ready to show the diff. '),
            ],
          },
          handled: {},
        },
      },
    },
  });
//
export const createShapeDiffMachine = (
  shapeTrail: IShapeTrail,
  diffs: ParsedDiff[],
  id: string,
  services: InteractiveSessionConfig
) =>
  Machine<
    DiffContext<IValueAffordanceSerializationWithCounter>,
    DiffStateSchema,
    DiffEvent
  >({
    id: id,
    context: {
      results: undefined,
      descriptionWhileLoading: descriptionForDiffs(diffs),
      preview: undefined,
    },
    initial: 'unfocused',
    states: {
      unfocused: {
        on: { SHOWING: 'loading' },
      },
      loading: {
        invoke: {
          id: 'loading-shape-diff-affordances',
          src: async (context, event) => {
            const firstDiff = diffs[0]!;
            const { pathId, method } = firstDiff.location();

            const trailValues = await services.diffService.learnTrailValues(
              services.rfcBaseState.rfcService,
              services.rfcBaseState.rfcId,
              pathId,
              method,
              firstDiff.raw()
            );

            return {
              preview: await prepareShapeDiffSuggestionPreview(
                diffs,
                services,
                trailValues
              ),
              results: trailValues,
            };
          },
          onDone: {
            target: 'ready',
            actions: assign({
              results: (context, event) => event.data.results, // cached in case we need to run 'em again
              preview: (context, event) => event.data.preview,
            }),
          },
        },
      },
      ready: {
        initial: 'unhandled',
        states: {
          unhandled: {
            entry: [(context, event) => console.log('ready to show the diff.')],
          },
          handled: {},
        },
      },
    },
  });
