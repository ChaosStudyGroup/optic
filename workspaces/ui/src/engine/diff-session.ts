import { assign, spawn, Machine, StateMachine, send } from 'xstate';
import { ParsedDiff } from './parse-diff';
import { DiffSet } from './diff-set';
import equals from 'lodash.isequal';
import { InteractiveSessionConfig } from './interfaces/session';
import {
  InteractiveEndpointSessionContext,
  InteractiveEndpointSessionEvent,
  newInteractiveEndpointSessionMachine,
} from './interactive-endpoint';
import { IUnrecognizedUrl } from '../services/diff';
import { IToDocument } from './interfaces/interfaces';

export interface DiffSessionSessionStateSchema {
  states: {
    loading: {};
    preparing: {};
    ready: {};
  };
}

// The events that the machine handles
export type DiffSessionSessionEvent =
  | { type: 'COMPLETED_DIFF'; diffs: ParsedDiff[]; urls: IUnrecognizedUrl[] }
  | { type: 'SHOWING' }
  | {
      type: 'SELECTED_ENDPOINT';
      pathId: string;
      method: string;
    }
  | {
      type: 'SEND_TO_ENDPOINT';
      pathId: string;
      method: string;
      event: InteractiveEndpointSessionEvent;
    }
  | { type: 'HANDLED_UPDATED'; pathId: string; method: string }
  | {
      type: 'TOGGLED_UNDOCUMENTED';
      active: boolean;
    }
  | {
      type: 'UPDATED_TO_DOCUMENT';
      handled: number;
      total: number;
      endpoints: { method: string; pathId: string; pathExpression: string }[];
      toDocument: IToDocument[];
    };

export function sendMessageToEndpoint(
  pathId: string,
  method: string,
  event: InteractiveEndpointSessionEvent
) {
  return {
    type: 'SEND_TO_ENDPOINT',
    pathId,
    method,
    event,
  };
}

export interface IEndpointWithStatus {
  pathId: string;
  method: string;
  handled: boolean;
  diffCount: number;
  machine: any;
  ref?: any;
}

// The context (extended state) of the machine
export interface DiffSessionSessionContext {
  allDiffs: ParsedDiff[];
  unrecognizedUrls: IUnrecognizedUrl[];
  endpoints: IEndpointWithStatus[];
  handledByEndpoint: IHandledByEndpoint;
  undocumentedEndpoints: { pathId: string; method: string }[];
  focus?: { pathId: string; method: string };
  showingUndocumented: boolean;
  unrecognizedUrlsToDocument: {
    urls: IToDocument[];
    endpoints: { pathId: string; method: string; pathExpression: string }[];
    handled: number;
    total: number;
  };
}

export const newDiffSessionSessionMachine = (
  diffId: string,
  services: InteractiveSessionConfig
) => {
  return Machine<
    DiffSessionSessionContext,
    DiffSessionSessionStateSchema,
    DiffSessionSessionEvent
  >({
    id: diffId,
    context: {
      handledByEndpoint: [],
      unrecognizedUrls: [],
      undocumentedEndpoints: [],
      allDiffs: [],
      endpoints: [],
      unrecognizedUrlsToDocument: {
        handled: 0,
        total: 0,
        endpoints: [],
        urls: [],
      },
      showingUndocumented: false,
    },
    initial: 'loading',
    states: {
      loading: {
        on: {
          COMPLETED_DIFF: {
            actions: assign({
              allDiffs: (context, event) => event.diffs,
              unrecognizedUrls: (context, event) => event.urls,
            }),
            target: 'preparing',
          },
        },
      },
      preparing: {
        invoke: {
          id: 'preparing-endpoints',
          src: async (context, event) => {
            const set = new DiffSet(context.allDiffs, services.rfcBaseState);

            const byEndpoint = set.groupedByEndpoint();
            const undocumented = set.forUndocumented();

            const undocumentedEndpoints = [];
            undocumented.iterator().forEach((i) => {
              const location = i.location(services.rfcBaseState);
              const entry = {
                method: location.method,
                pathId: location.pathId,
              };
              if (!undocumentedEndpoints.find((i) => equals(i, entry)))
                undocumentedEndpoints.push(entry);
            });

            const endpoints: IEndpointWithStatus[] = byEndpoint.map(
              ({ pathId, method, diffs }) => {
                return {
                  pathId,
                  handled: false,
                  diffCount: diffs.length,
                  method,
                  machine: newInteractiveEndpointSessionMachine(
                    pathId,
                    method,
                    diffs,
                    services
                  ),
                };
              }
            );
            return { endpoints, undocumentedEndpoints };
          },
          onDone: {
            target: 'ready',
            actions: [
              assign({
                undocumentedEndpoints: (_, event) =>
                  event.data.undocumentedEndpoints,
                endpoints: (context, event) => {
                  return event.data.endpoints.map((endpoint) => {
                    return {
                      ...endpoint,
                      ref: spawn(
                        endpoint.machine,
                        `${endpoint.pathId}${endpoint.method}`
                      ),
                    };
                  });
                },
              }),
            ],
          },
        },
      },
      ready: {
        on: {
          HANDLED_UPDATED: {
            actions: [
              assign({
                handledByEndpoint: (context) => computeHandled(context),
              }),
            ],
          },
          SEND_TO_ENDPOINT: {
            actions: (context, event) => {
              const actor = context.endpoints.find(
                (i) => i.pathId === event.pathId && i.method === event.method
              );
              if (actor) {
                actor.ref.send(event.event);
              }
            },
          },
          TOGGLED_UNDOCUMENTED: {
            actions: assign({
              showingUndocumented: (context, event) => event.active,
              focus: (ctx, event) => undefined,
            }),
          },
          UPDATED_TO_DOCUMENT: {
            actions: assign({
              unrecognizedUrlsToDocument: (context, event) => ({
                urls: event.toDocument,
                endpoints: event.endpoints,
                handled: event.handled,
                total: event.total,
              }),
            }),
          },
          SELECTED_ENDPOINT: {
            actions: [
              assign({
                focus: (context, event) => ({
                  pathId: event.pathId,
                  method: event.method,
                }),
                showingUndocumented: (ctx, event) => false,
              }),
              (context) => {
                const endpointMachine = context.endpoints.find(
                  (i) =>
                    i.pathId === context.focus.pathId &&
                    i.method === context.focus.method
                )!.ref;
                send({ type: 'PREPARE' }, { to: endpointMachine }); //warm it up
              },
            ],
          },
        },
      },
    },
  });
};

type IHandledByEndpoint = {
  pathId: string;
  method: string;
  diffCount: number;
  handled: number;
}[];
function computeHandled(
  context: DiffSessionSessionContext
): IHandledByEndpoint {
  const results: IHandledByEndpoint = [];

  context.endpoints.forEach((i) => {
    const endpointContext: InteractiveEndpointSessionContext =
      i.ref.state.context;
    const entries: [string, boolean][] = Object.entries(
      endpointContext.handledByDiffHash
    );
    const handled = entries.filter(([key, bool]) => bool).length;
    results.push({
      pathId: i.pathId,
      method: i.method,
      handled,
      diffCount: entries.length !== 0 ? entries.length : i.diffCount,
    });
  });

  return results;
}

export function nextEndpointToFocusOn(
  context: DiffSessionSessionContext,
  endpointsWithDiffs: { pathId: string; method: string }[]
) {
  const indexOfCurrentFoucs: number =
    context.focus &&
    endpointsWithDiffs.findIndex(
      ({ pathId, method }) =>
        pathId === context.focus.pathId && method === context.focus.method
    );

  function alreadyHandled(pathId: string, method: string) {
    const status = context.handledByEndpoint.find(
      (i) => pathId === i.pathId && method === i.method
    );
    return (
      status && status.diffCount > 0 && status.diffCount === status.handled
    );
  }
  // next unhandled after this one, or start from top of list.
  const searchOrder = [
    ...endpointsWithDiffs.slice(
      indexOfCurrentFoucs || 0,
      endpointsWithDiffs.length
    ),
    ...endpointsWithDiffs.slice(0, indexOfCurrentFoucs || 0),
  ].filter(
    (i) =>
      !(
        i.method === context.focus.method && i.pathId === context.focus.pathId
      ) && !alreadyHandled(i.pathId, i.method)
  );

  const next = searchOrder[0];

  return next ? { pathId: next.pathId, method: next.method } : context.focus;
}
