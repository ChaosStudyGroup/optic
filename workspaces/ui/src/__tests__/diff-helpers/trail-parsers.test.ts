import { makeUniverse } from './universes/makeUniverse';
import { makeDiffRfcBaseState } from '../../engine/interfaces/diff-rfc-base-state';
import { ParsedDiff } from '../../engine/parse-diff';

const universe_raw = require('./universes/simple-todo/universe.json');

const universePromise = makeUniverse(universe_raw);

test('accurate spec trail for all diffs', async () => {
  const { captureService, diffService, rfcBaseState } = await universePromise;
  const diffsRaw = (await diffService.listDiffs()).rawDiffs;

  const parsedDiffs = diffsRaw.map(
    ([diff, interactions]) => new ParsedDiff(diff, interactions)
  );

  parsedDiffs.forEach((i) =>
    expect(i.location(rfcBaseState)).toMatchSnapshot()
  );
});

// makeDiffRfcBaseState
