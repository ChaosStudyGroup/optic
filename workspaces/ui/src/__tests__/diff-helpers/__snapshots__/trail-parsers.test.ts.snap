// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accurate spec trail for all diffs 1`] = `
Object {
  "inRequest": Object {
    "contentType": "application/json",
  },
  "inResponse": false,
  "method": "GET",
  "pathId": "path_UOIsxzICu5",
}
`;

exports[`accurate spec trail for all diffs 2`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "GET",
  "pathId": "path_it2OyjUysW",
}
`;

exports[`accurate spec trail for all diffs 3`] = `
Object {
  "inRequest": false,
  "inResponse": Object {
    "contentType": "application/json",
    "statusCode": 200,
  },
  "method": "GET",
  "pathId": "path_xhUZ8irdJO",
}
`;

exports[`accurate spec trail for all diffs 4`] = `
Object {
  "inRequest": Object {
    "contentType": "application/json",
  },
  "inResponse": false,
  "method": "GET",
  "pathId": "path_xhUZ8irdJO",
}
`;

exports[`accurate spec trail for all diffs 5`] = `
Object {
  "inRequest": false,
  "inResponse": Object {
    "contentType": "application/json",
    "statusCode": 200,
  },
  "method": "GET",
  "pathId": "path_wPHNu8BDab",
}
`;

exports[`accurate spec trail for all diffs 6`] = `
Object {
  "inRequest": Object {
    "contentType": "application/json",
  },
  "inResponse": false,
  "method": "GET",
  "pathId": "path_wPHNu8BDab",
}
`;

exports[`accurate spec trail for all diffs 7`] = `
Object {
  "inRequest": false,
  "inResponse": Object {
    "contentType": "application/json",
    "statusCode": 200,
  },
  "method": "GET",
  "pathId": "path_UOIsxzICu5",
}
`;

exports[`accurate spec trail for all diffs 8`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "GET",
  "pathId": "path_AsEexQkVwC",
}
`;

exports[`accurate spec trail for all diffs 9`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "POST",
  "pathId": "path_UOIsxzICu5",
}
`;

exports[`accurate spec trail for all diffs 10`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "POST",
  "pathId": "path_UOIsxzICu5",
}
`;

exports[`accurate spec trail for all diffs 11`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "GET",
  "pathId": "path_AsEexQkVwC",
}
`;

exports[`accurate spec trail for all diffs 12`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "GET",
  "pathId": "path_it2OyjUysW",
}
`;

exports[`accurate spec trail for all diffs 13`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "PATCH",
  "pathId": "path_it2OyjUysW",
}
`;

exports[`accurate spec trail for all diffs 14`] = `
Object {
  "inRequest": false,
  "inResponse": false,
  "method": "PATCH",
  "pathId": "path_it2OyjUysW",
}
`;
